virtualenv_embedded/distutils-init.py

```diff
@@ -15,7 +14,29 @@
     warnings.warn("The virtualenv distutils package at %s appears to be in the same location as the system distutils?")
 else:
     __path__.insert(0, distutils_path)  # noqa: F821
-    real_distutils = imp.load_module("_virtualenv_distutils", None, distutils_path, ("", "", imp.PKG_DIRECTORY))
+    if sys.version_info < (3, 3):
+        import imp
+
+        real_distutils = imp.load_module("_virtualenv_distutils", None, distutils_path, ("", "", imp.PKG_DIRECTORY))
+    else:
+        import importlib.machinery
+
+        distutils_path = os.path.join(distutils_path, "__init__.py")
+        loader = importlib.machinery.SourceFileLoader("_virtualenv_distutils", distutils_path)
+        if sys.version_info < (3, 4):
+            real_distutils = loader.load_module()
+        else:
+            if sys.version_info < (3, 5):
+                import types
+
+                real_distutils = types.ModuleType(loader.name)
+            else:
+                import importlib.util
+
+                spec = importlib.util.spec_from_loader(loader.name, loader)
+                real_distutils = importlib.util.module_from_spec(spec)
+            loader.exec_module(real_distutils)
```

Comment:
can we add a test that exposes this, aka somehow assert that no warnings raised during virtualenv run?

Reply:
I think the test would be `venv/bin/python -Werror -c 'import distutils'`?

Reply:
yeah to test just this, in more general would ``python -W -m virtualenv`` work 🤔?

Reply:
the warning used to happen in a subprocess during virtualenv so probably not I'm afraid

Reply:
we can maybe have some internal dev flags that would allow us to run with -Werror in all subprocesses 🤔

Reply:
oh, or `PYTHONWARNINGS=error ...` would work

Reply:
hahaha nope, can't do it until pip upgrades vendored requests:

```
=================================== FAILURES ===================================
____________________________ test_commandline_basic ____________________________

tmpdir = local('/tmp/pytest-of-asottile/pytest-24/test_commandline_basic0')

    def test_commandline_basic(tmpdir):
        """Simple command line usage should work and files should be generated"""
        home_dir, lib_dir, inc_dir, bin_dir = virtualenv.path_locations(str(tmpdir.join("venv")))
        cmd = [sys.executable, VIRTUALENV_SCRIPT, home_dir]
        env = dict(os.environ, PYTHONWARNINGS=str('error'))
>       subprocess.check_call(cmd, env=env)

tests/test_cmdline.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

popenargs = (['/tmp/virtualenv/.tox/py36/bin/python', '/tmp/virtualenv/virtualenv.py', '/tmp/pytest-of-asottile/pytest-24/test_commandline_basic0/venv'],)
kwargs = {'env': {'COVERAGE_FILE': '/tmp/virtualenv/.tox/.coverage.py36', 'HOME': '/home/asottile', 'LANG': 'en_US.UTF-8', 'PAT...e/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin', ...}}
retcode = 1
cmd = ['/tmp/virtualenv/.tox/py36/bin/python', '/tmp/virtualenv/virtualenv.py', '/tmp/pytest-of-asottile/pytest-24/test_commandline_basic0/venv']

    def check_call(*popenargs, **kwargs):
        """Run command with arguments.  Wait for command to complete.  If
        the exit code was zero then return, otherwise raise
        CalledProcessError.  The CalledProcessError object will have the
        return code in the returncode attribute.
    
        The arguments are the same as for the call function.  Example:
    
        check_call(["ls", "-l"])
        """
        retcode = call(*popenargs, **kwargs)
        if retcode:
            cmd = kwargs.get("args")
            if cmd is None:
                cmd = popenargs[0]
>           raise CalledProcessError(retcode, cmd)
E           subprocess.CalledProcessError: Command '['/tmp/virtualenv/.tox/py36/bin/python', '/tmp/virtualenv/virtualenv.py', '/tmp/pytest-of-asottile/pytest-24/test_commandline_basic0/venv']' returned non-zero exit status 1.

/usr/lib/python3.6/subprocess.py:291: CalledProcessError
----------------------------- Captured stdout call -----------------------------
Using real prefix '/usr'
Path not in prefix '/tmp/virtualenv/.tox/py36/include/python3.6m' '/usr'
New python executable in /tmp/pytest-of-asottile/pytest-24/test_commandline_basic0/venv/bin/python
Complete output from command /tmp/pytest-of-asott...sic0/venv/bin/python -m pip config list:
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/__main__.py", line 16, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/__init__.py", line 40, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/cli/autocompletion.py", line 8, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/cli/main_parser.py", line 12, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/commands/__init__.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/commands/completion.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/cli/base_command.py", line 18, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_internal/download.py", line 15, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_vendor/requests/__init__.py", line 116, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_vendor/requests/models.py", line 43, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 951, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 894, in _find_spec
  File "<frozen importlib._bootstrap_external>", line 1157, in find_spec
  File "<frozen importlib._bootstrap_external>", line 1131, in _get_spec
  File "<frozen importlib._bootstrap_external>", line 1112, in _legacy_get_spec
  File "<frozen importlib._bootstrap>", line 441, in spec_from_loader
  File "<frozen importlib._bootstrap_external>", line 544, in spec_from_file_location
  File "/tmp/virtualenv/virtualenv_support/pip-18.1-py2.py3-none-any.whl/pip/_vendor/requests/status_codes.py", line 18
SyntaxError: invalid escape sequence \o
----------------------------------------
----------------------------- Captured stderr call -----------------------------
Traceback (most recent call last):
  File "/tmp/virtualenv/virtualenv.py", line 2466, in <module>
    main()
  File "/tmp/virtualenv/virtualenv.py", line 769, in main
    symlink=options.symlink,
  File "/tmp/virtualenv/virtualenv.py", line 1053, in create_environment
    install_wheel(to_install, py_executable, search_dirs, download=download)
  File "/tmp/virtualenv/virtualenv.py", line 930, in install_wheel
    config = _pip_config(py_executable, python_path)
  File "/tmp/virtualenv/virtualenv.py", line 1004, in _pip_config
    remove_from_env=["PIP_VERBOSE", "PIP_QUIET"],
  File "/tmp/virtualenv/virtualenv.py", line 862, in call_subprocess
    raise OSError("Command {} failed with error code {}".format(cmd_desc, proc.returncode))
OSError: Command /tmp/pytest-of-asott...sic0/venv/bin/python -m pip config list failed with error code 1
============== 1 failed, 2 passed, 32 deselected in 6.18 seconds ===============
ERROR: InvocationError for command '/tmp/virtualenv/.tox/py36/bin/coverage run --source=virtualenv -m pytest tests -k test_cmdline' (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   py36: commands failed
```

Reply:
Here's the patch for that though:

```diff
diff --git a/tests/test_cmdline.py b/tests/test_cmdline.py
index b8b357d..9aa260c 100644
--- a/tests/test_cmdline.py
+++ b/tests/test_cmdline.py
@@ -22,7 +22,9 @@ VIRTUALENV_SCRIPT = get_src(virtualenv.__file__)
 def test_commandline_basic(tmpdir):
     """Simple command line usage should work and files should be generated"""
     home_dir, lib_dir, inc_dir, bin_dir = virtualenv.path_locations(str(tmpdir.join("venv")))
-    subprocess.check_call([sys.executable, VIRTUALENV_SCRIPT, home_dir])
+    cmd = [sys.executable, VIRTUALENV_SCRIPT, home_dir]
+    env = dict(os.environ, PYTHONWARNINGS=str('error'))
+    subprocess.check_call(cmd, env=env)
 
     assert os.path.exists(home_dir)
     assert os.path.exists(bin_dir)
```

Reply:
let's add it with expected fail then 👍otherwise we'll lose it here as a comment